### ISSUE TITLE
A case where @CalledMethods doesn't get all methods being called 

### ISSUE THREAD

**Issue by ranmanli** [kelloggm/object-construction-checker#42]
I was running this checker in the gapic-generator (https://github.com/googleapis/gapic-generator). In PackageMetadataConfig.java:303 it reported warning as:

```
/gapic-generator/src/main/java/com/google/api/codegen/config/PackageMetadataConfig.java:303: error: [method.invocation.invalid] call to build() not allowed on the given receiver.
    return builder.build();
                        ^
  found   : @CalledMethods({"apiCommonVersionBound", "authVersionBound", "commonsCliVersionBound", "gaxGrpcVersionBound", "gaxHttpVersionBound", "gaxVersionBound", "generatedNonGAPackageVersionBound", "grpcVersionBound", "packageName", "protoVersionBound", "releaseLevel"}) Builder
  required: @CalledMethods({"apiCommonVersionBound", "apiVersion", "authVersionBound", "author", "commonsCliVersionBound", "email", "gaxGrpcVersionBound", "gaxHttpVersionBound", "gaxVersionBound", "generatedNonGAPackageVersionBound", "grpcVersionBound", "homepage", "licenseName", "packageName", "protoPackageDependencies", "protoPath", "protoVersionBound", "releaseLevel", "shortName"}) Builder
```

It indicates methods "author", "email", "homepage", "liscenseName", "protoPackageDependencies", "protoPath", "shortName" are not called by the builder which is not true refer to the code. 

```java
  public static PackageMetadataConfig createFromPackaging(
      ApiDefaultsConfig apiDefaultsConfig,
      DependenciesConfig dependenciesConfig,
      PackagingConfig packagingConfig) {
    // dependencies config
    Builder builder =
        newBuilder()
            .gaxVersionBound(dependenciesConfig.gaxVersionBound())
            .gaxGrpcVersionBound(dependenciesConfig.gaxGrpcVersionBound())
            .gaxHttpVersionBound(dependenciesConfig.gaxHttpVersionBound())
            .grpcVersionBound(dependenciesConfig.grpcVersionBound())
            .protoVersionBound(dependenciesConfig.protoVersionBound())
            .authVersionBound(dependenciesConfig.authVersionBound())
            .apiCommonVersionBound(dependenciesConfig.apiCommonVersionBound())
            .commonsCliVersionBound(dependenciesConfig.commonsCliVersionBound());

    // api_defaults config
    builder
 .generatedNonGAPackageVersionBound(apiDefaultsConfig.generatedNonGAPackageVersionBound())
        .author(apiDefaultsConfig.author())
        .email(apiDefaultsConfig.email())
        .homepage(apiDefaultsConfig.homepage())
        .licenseName(apiDefaultsConfig.licenseName());

    // packaging config
    builder.releaseLevel(apiDefaultsConfig.releaseLevel());
    if (packagingConfig != null) {
      builder
          .packageName(packagingConfig.packageName())
          .shortName(packagingConfig.apiName())
          .artifactType(packagingConfig.artifactType())
          .apiVersion(packagingConfig.apiVersion())
          .protoPath(packagingConfig.protoPath());

      // config depending on multiple sources
      Map<TargetLanguage, Map<String, VersionBound>> protoPackageDependencies =
          createProtoPackageDependencies(
              packagingConfig.protoPackageDependencies(), dependenciesConfig);
      builder.protoPackageDependencies(protoPackageDependencies);

      Map<TargetLanguage, Map<String, VersionBound>> protoPackageTestDependencies =
          createProtoPackageDependencies(
              packagingConfig.protoPackageTestDependencies(), dependenciesConfig);
      builder.protoPackageTestDependencies(protoPackageTestDependencies);
      if (packagingConfig.releaseLevel() != null) {
        builder.releaseLevel(packagingConfig.releaseLevel());
      }
    } else {
      builder
          .packageName("")
          .shortName("")
          .artifactType(PackagingArtifactType.GAPIC) // the most likely case
          .apiVersion("")
          .protoPath("")
          .protoPackageDependencies(ImmutableMap.of())
          .protoPackageTestDependencies(ImmutableMap.of());
    }

    return builder.build();
  }
```

**Comment by msridhar** [comment]
At some point we need a custom error message just indicating which methods you forgot to call...

This is a weird one.  `gaxGrpcVersionBound` is called right after the builder is created but somehow we miss it.  I added this code locally to the `Animal.java` AutoValue test and it works fine:

```java
  public static void buildSomethingRightFluentWithLocal() {
    Builder b = builder().setName("Jim").setNumberOfLegs(7);
    b.build();
  }
```

@kelloggm any idea what might be happening here?  I'll check if the returns-receiver checker is injecting `@This` appropriately.

**Comment by kelloggm** [comment]
Here's a minimal test case (`Issue42.java`):

```java
import com.google.auto.value.AutoValue;
import org.checkerframework.checker.builder.qual.*;
import org.checkerframework.checker.nullness.qual.*;

/**
 * Adapted from the standard AutoValue example code:
 * https://github.com/google/auto/blob/master/value/userguide/builders.md
 */
@AutoValue
abstract class Issue42 {
  abstract String name();

  abstract @Nullable String habitat();

  abstract int numberOfLegs();

  abstract String anotherProperty();

  public String getStr() {
    return "str";
  }

  public abstract Builder toBuilder();

  static Builder builder() {
    return new AutoValue_Issue42.Builder();
  }

  @AutoValue.Builder
  abstract static class Builder {

    abstract Builder setName(String value);

    abstract Builder setNumberOfLegs(int value);

    abstract Builder setHabitat(String value);

    abstract Builder setAnotherProperty(String value);

    abstract Issue42 build();
  }

  public static void buildSomethingWrong(Object packagingConfig) {
    Builder builder =
            builder().setName("Frank");

    // more config
    builder.setHabitat("Airplane");

    builder.setNumberOfLegs(5);
    if (packagingConfig != null) {
      builder.setAnotherProperty("foo");

      if (packagingConfig.toString() != null) {
        builder.setNumberOfLegs(3);
      }
    } else {
      builder.setAnotherProperty("bar");
    }
    return builder.build();
  }
}
```

**Comment by kelloggm** [comment]
That issues an error on the return statement at line 60, which I think is the one we wanted.

**Comment by msridhar** [comment]
Nice reduction, @kelloggm!  Are you looking into the root cause?  If not I can try.

**Comment by kelloggm** [comment]
I'll look into the root cause. I had another meeting I had to take, sorry.

**Comment by kelloggm** [comment]
@msridhar I realized that the error that was being issued wasn't the same one that @ranmanli reported. The test, when corrected, does not issue an error; sorry for misleading you. I tried adding in everything I could imagine might have an impact from the original test case posted by @ranmanli but I cannot reproduce. Here is the updated test case I was using, which does not issue any errors:

```java
import com.google.auto.value.AutoValue;
import org.checkerframework.checker.builder.qual.*;
import org.checkerframework.checker.nullness.qual.*;
import java.util.Map;

/**
 * Adapted from the standard AutoValue example code:
 * https://github.com/google/auto/blob/master/value/userguide/builders.md
 */
@AutoValue
abstract class Issue42 {
  abstract String name();

  abstract String habitat();

  abstract int numberOfLegs();

  abstract String anotherProperty();

  abstract @Nullable Map<String, String> map();

  public String getStr() {
    return "str";
  }

  public abstract Builder toBuilder();

  static Builder builder() {
    return new AutoValue_Issue42.Builder();
  }

  @AutoValue.Builder
  abstract static class Builder {

    abstract Builder setName(String value);

    abstract Builder setNumberOfLegs(int value);

    abstract Builder setHabitat(String value);

    abstract Builder setAnotherProperty(String value);

    abstract Builder setMap(Map<String, String> value);

    abstract Issue42 build();
  }

  public static Map<String, String> createMap(Object packagingConfig) {
    return null;
  }

  public static Issue42 buildSomething(Object packagingConfig) {
    Builder builder =
            builder().setName("Frank");

    // more config
    builder.setHabitat("Airplane");

    builder.setNumberOfLegs(5);
    if (packagingConfig != null) {
      builder.setAnotherProperty("foo");

      Map<String, String> myMap =
              createMap(packagingConfig);
      builder.setMap(myMap);

      if (packagingConfig.toString() != null) {
        builder.setNumberOfLegs(3);
      }
    } else {
      builder.setAnotherProperty("bar");
    }
    return builder.build();
  }
}
```

**Comment by msridhar** [comment]
Ok I will take a look.

**Comment by msridhar** [comment]
Found it!  Here's the test case:

```java
  public static void buildSomethingRightFluentWithLocal() {
    Builder b = builder();
    b.setName("Jim").setNumberOfLegs(7);
    b.build();
  }
```

I've pushed it on the `nested-receiver-bug` branch.  The issue here is that for the statement `b.setName("Jim").setNumberOfLegs(7);`, our transfer function in `TypesafeBuilderTransfer` updates the type of `b.setName("Jim")` to be `@CalledMethods({"setName","setNumberOfLegs"})`, but it does *not* also update the type of `b`.

This is a nasty one.  @kelloggm I guess maybe we need to recurse in the transfer function, depending on whether we have `@This` types on nested receivers?  Not sure of the best fix.

**Comment by kelloggm** [comment]
Nice work! I'll see if I can figure out a fix.
